Refactored metrics_df in report_model_metrics for aesthetics

lshpaner opened the pull request:
Previously, `metrics_df` was printing out all of the metrics on one line, making the output look less than aesthetically pleasing, and, more importantly, difficult to follow; See issue below:

https://github.com/uclamii/model_tuner/issues/117

This became even more cumbersome and problematic when dealing with multi-class output; see example below:

```bash
   0 Precision/PPV  0 Sensitivity/Recall  0 F1-Score  1 Precision/PPV  \
0         0.833333                   1.0    0.909091         0.888889   

   1 Sensitivity/Recall  1 F1-Score  2 Precision/PPV  2 Sensitivity/Recall  \
0                   0.8    0.842105              1.0                   0.9   

   2 F1-Score  macro avg Precision/PPV  macro avg Sensitivity/Recall  \
0    0.947368                 0.907407                           0.9   

   macro avg F1-Score  weighted avg Precision/PPV  \
0            0.899522                    0.907407   

   weighted avg Sensitivity/Recall  weighted avg F1-Score  \
0                              0.9               0.899522   

   Weighted Average Precision  Weighted Average Recall  Multiclass AUC ROC  
0                    0.907407                      0.9            0.933333  
```

This pull request fixes the `metrics_df` on one line, before the return statement as follows:

Before:

```python
metrics_df = pd.DataFrame(metrics, index=[0])
```

After:

```python
metrics_df = pd.DataFrame(metrics, index=[0]).T.rename(columns={0: ""})
```


```bash
                                         
0 Precision/PPV                  0.833333
0 Sensitivity/Recall             1.000000
0 F1-Score                       0.909091
1 Precision/PPV                  0.888889
1 Sensitivity/Recall             0.800000
1 F1-Score                       0.842105
2 Precision/PPV                  1.000000
2 Sensitivity/Recall             0.900000
2 F1-Score                       0.947368
macro avg Precision/PPV          0.907407
macro avg Sensitivity/Recall     0.900000
macro avg F1-Score               0.899522
weighted avg Precision/PPV       0.907407
weighted avg Sensitivity/Recall  0.900000
weighted avg F1-Score            0.899522
Weighted Average Precision       0.907407
Weighted Average Recall          0.900000
Multiclass AUC ROC               0.933333
```

## Updates - 12/10/2024

### Updated `report_model_metrics` to:

- output `class_labels` as they are provided in the `Model`
- included print of `report_model_metrics` set to `True` by default
- made `return_dict` optional input in `return_metrics`

### Updated `xgb_early_test.py` script:
  - renamed to `xgb_early_smote_test.py` b/c it has `SMOTE` in it
  - added `report_model_metrics` to test this method on the script
  - added `random_state` to SMOTE(); otherwise, results are not reproducible; will be mentioning this in docs as well

```bash
--------------------------------------------------------------------------------
Setosa Precision/PPV: 0.8333
Setosa Sensitivity/Recall: 1.0000
Setosa F1-Score: 0.9091
--------------------------------------------------------------------------------
Versicolor Precision/PPV: 0.8889
Versicolor Sensitivity/Recall: 0.8000
Versicolor F1-Score: 0.8421
--------------------------------------------------------------------------------
Virginica Precision/PPV: 1.0000
Virginica Sensitivity/Recall: 0.9000
Virginica F1-Score: 0.9474
--------------------------------------------------------------------------------
macro avg Precision/PPV: 0.9074
macro avg Sensitivity/Recall: 0.9000
macro avg F1-Score: 0.8995
--------------------------------------------------------------------------------
weighted avg Precision/PPV: 0.9074
weighted avg Sensitivity/Recall: 0.9000
weighted avg F1-Score: 0.8995
--------------------------------------------------------------------------------
Weighted Average Precision: 0.9074
Weighted Average Recall: 0.9000
Multiclass AUC ROC: 0.9333
--------------------------------------------------------------------------------
********************************************************************************
```


Additionally:

- updated R^2 print out to R²

```bash
Validation Metrics
********************************************************************************
R²: 0.5899
Explained Variance: 0.6139
Mean Absolute Error: 8753152.1062
Median Absolute Error: 4129556.0000
Mean Squared Error: 220513062869269.0938
RMSE: 14849682.2481
********************************************************************************
Test Metrics
********************************************************************************
R²: 0.1567
Explained Variance: 0.1570
Mean Absolute Error: 5399847.9062
Median Absolute Error: 2463762.1250
Mean Squared Error: 76866085832469.8750
RMSE: 8767330.5990
********************************************************************************
Report model metrics for XGB
Mean Absolute Error (MAE): 5399847.9062
Mean Squared Error (MSE): 76866085832469.8750
Root Mean Squared Error (RMSE): 8767330.5990
R² Score: 0.1567
Explained Variance: 0.1570
********************************************************************************
```

panas89 commented:
regression needed the following library openpyxl
need to define corresponding version for each python version, in requierements

after installing the latest could reproduce all ur results on ur branch

lshpaner commented:
> regression needed the following library openpyxl need to define corresponding version for each python version, in requierements
> 
> after installing the latest could reproduce all ur results on ur branch

Done

lshpaner commented:
## Update - 12/11/2024

This additional PR:

- Introduces a `return_dict` parameter to the `return_metrics` method, providing users with the flexibility to optionally return evaluation metrics as a dictionary while preserving the original printed output.
- Adds a docstring to the `return_metrics` method
- Adds direct print inside `report_model_metrics` method that specifies the title `Report Model Metrics`, followed by the dynamic `estimator_name` from the `Model`


### Add `return_dict` Parameter to `return_metrics` Function

---

#### Changes Made
1. **Preserved Original Print Outputs**:
   - The function retains all existing print logic for confusion matrices, classification reports, and additional metrics.

2. **Optional Dictionary Return**:
   - Added a `return_dict` parameter (default `False`) to enable returning a dictionary containing metrics like:
     - `Classification Report`
     - `Confusion Matrix`
     - `Best Features` (if `feature_selection` is enabled)
     - `Regression Report` (for regression models)

3. **Improved Usability**:
   - The function now seamlessly integrates dictionary-based outputs without altering its primary behavior, ensuring backward compatibility.

---

 
#### Usage Examples
1. **Default Behavior**:
   ```python
   model.return_metrics(X=X_valid, y=y_valid)
   # Prints metrics but does not return anything
   ```

2. **Returning Metrics as a Dictionary**:
   ```python
   results = model.return_metrics(X=X_valid, y=y_valid, return_dict=True)
   print(results["Classification Report"])
   ```

---

## Update - 12/12/2024

- Fixed K Fold metrics bug:

  ```bash
  ValueError: The `test_model` must have a `model_type` attribute.
  ```
  - Ensured that the necessary attributes were in place inside `return_metrics_kfold` for printing `Return Model Metrics: [estimator_name]`
  - added thees lines in`return_metrics_kfold`: 

    ```python
    # Ensure test_model has necessary attributes
    if not hasattr(test_model, "model_type"):
        test_model.model_type = self.model_type
    if not hasattr(test_model, "estimator_name"):
        test_model.estimator_name = self.estimator_name  # Pass from the main model
    ```
